File: ChibuzorTasks/bankapp/src/bankapp/Bank.java
package bankapp.src.bankapp;


import java.util.ArrayList;

public class Bank {
    private ArrayList<Account> accounts;
    private String bankName;
    private int accountCounter;


    public Bank( String bankName ) {
        this.bankName = bankName;
        this.accountCounter = 0;
        this.accounts = new ArrayList<>();

    }

    public void createAccount(String firstName, String lastName, String pin) {
        if (firstName == null || lastName == null || pin == null) {
            throw new IllegalArgumentException("First name, last name, and PIN cannot be null");
        }
        if (firstName.isEmpty() || lastName.isEmpty() || pin.isEmpty()) {
            throw new IllegalArgumentException("First name, last name, and PIN cannot be empty");
        }
        accounts.add(new Account(firstName, lastName, generateAccountNumber(), pin));
    }

    public String generateAccountNumber() {
        String number = String.valueOf(accountCounter);
        accountCounter++;
        return number;

    }

    public void transfer(String sender, String recipient, double amount, String pin) {
        Account senderAccount = findAccount(sender);
        Account recipientAccount = findAccount(recipient);
        if (senderAccount == null || recipientAccount == null) throw new IllegalArgumentException("Sender or recipient not found");
        if (senderAccount == recipientAccount ) throw new IllegalArgumentException("Sender and recipient can not be same");
        senderAccount.withdraw(amount, pin);
        recipientAccount.deposit(amount);
    }


    private Account findAccount(String accountNumber) {
        if (accountNumber == null || accountNumber.isEmpty()) throw new IllegalArgumentException("Invalid account number");
        if (accounts.isEmpty()) throw new IllegalArgumentException("No account found");
        for (Account account : accounts) {
            if (account.getAccountNumber().equals(accountNumber)) return account;
        }
        return null;
    }

    public ArrayList<Account> getAccounts() {
        return accounts;
    }

    public String getBankName() {
        return bankName;
    }

    public void deposit(String accountNumber, double amount) {
        Account account = findAccount(accountNumber);
        if (account == null) throw new IllegalArgumentException();
        if (amount < 0) throw new IllegalArgumentException();
        account.deposit(amount);
    }

    public double checkBalance(String accountNumber, String pin) {
        if (accountNumber == null) throw new IllegalArgumentException("Account cannot be null");
        if (accounts.isEmpty()) throw new IllegalArgumentException("No account found");
        if (pin == null) throw new IllegalArgumentException("Invalid pin");
        Account account = findAccount(accountNumber);
        if (account == null) throw new IllegalArgumentException("No account found");
        return account.getBalance();
    }
}

File: PythonClass/classWorks/Feb-2025/mybank/bank.py
from account import Account

class Bank:
    def __init__(self, bank_name):
        self.bank_name = bank_name
        self.account_counter = 0
        self.accounts = []

    def create_account(self, first_name, last_name, pin):
        if not first_name or not last_name or not pin:
            raise ValueError("First name, last name, and PIN cannot be null or empty")
        account = Account(first_name, last_name, self.generate_account_number(), pin)
        self.accounts.append(account)

    def generate_account_number(self):
        number = str(self.account_counter)
        self.account_counter += 1
        return number

    def transfer(self, sender, recipient, amount, pin):
        sender_account = self.find_account(sender)
        recipient_account = self.find_account(recipient)
        if not sender_account or not recipient_account:
            raise ValueError("Sender or recipient not found")
        if sender_account == recipient_account:
            raise ValueError("Sender and recipient cannot be the same")
        sender_account.withdraw(amount, pin)
        recipient_account.deposit(amount)

    def find_account(self, account_number):
        if not account_number:
            raise ValueError("Invalid account number")
        for account in self.accounts:
            if account.get_account_number() == account_number:
                return account
        return None

    def get_accounts(self):
        return self.accounts

    def get_bank_name(self):
        return self.bank_name

    def deposit(self, account_number, amount):
        account = self.find_account(account_number)
        if not account:
            raise ValueError("Account not found")
        if amount < 0:
            raise ValueError("Amount must be positive")
        account.deposit(amount)

    def check_balance(self, account_number, pin):
        if not account_number:
            raise ValueError("Account number cannot be null")
        if not pin:
            raise ValueError("Invalid pin")

        account = self.find_account(account_number)
        if not account:
            raise ValueError("No account found")
        if account.get_pin() != pin:
            raise ValueError("Invalid PIN")

        return account.get_balance()

